### PR TITLE
Env: Only perform expensive install work when required

### PR DESCRIPTION
**Pull request by noahtallen**
## Description

TLDR: wp-env start only takes 1-2 seconds in common use cases.

`wp-env start` is now much lazier, only doing time-intensive work when required.

`wp-env` previously did the following actions on every run:

- WordPress configuration is applied (e.g. `wp core install` and `wp config set`)
- Download specified remote sources like WordPress
- Ran wp-env stop as a "restart" mechanism to handle some edge cases

Now, these actions are only performed in two situations:
1. The parsed wp-env configuration is different. (Configuration is considered different on first install, when updating `*.wp-env.json` files, and when different environment variables change.)
2. The user specifies the `--update` flag.

To accomplish this, I added a little library to read/write cache values to a cache JSON file stored in the wp-env work directory. This might be useful in the future, particularly if we want to get more granular about this behavior.

_Note that the WordPress install and source downloads still happen each run so that they stay up-to-date._

## How has this been tested?
Locally, in wp-env; CI should work. You can test with `npx wp-env start` locally with this branch checked out and it will run sourcecode.

## Screenshots 

The first run is before this change. The second is with this improvement:

Screen Shot 2020-07-08 at 6 20 12 PM

## Types of changes
Performance improvement

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/env/lib/commands/start.js [hunk 1]
```diff
@@ -26,6 +26,8 @@ const {
 	configureWordPress,
 	setupWordPressDirectories,
 } = require( '../wordpress' );
+const { didMetaChange, setMeta } = require( '../meta' );
```
Comment (epiqueras) [inline packages/env/lib/commands/start.js:29]: This naming kind of obscures what this is. It's not a metafile; it's more like a cache file or a hash list.

Comment (noahtallen) [reply packages/env/lib/commands/start.js:29]: I tried to deliberately generalize it into a metafile -- we can `setMeta` to whatever value you want and then retrieve that same value later. In the case of `start` that value happens to be a hash of the config file

But I'm happy to change the name to something more like "cache". That could work in both ways, and also drives home the point that it's "temporary" (i.e. everything should still work if the cache is not set)

**Comment by noahtallen** [comment]
@epiqueras do you know if we need run the WordPress configuration code again if the WordPress source updates? Currently, if it does a `git pull` (or whatever) on the WordPress source, it will not run install/configuration again. (it only does that if the config is new to your system or has changed)

**Comment by epiqueras** [comment]
I guess it depends on the nature of the changes that have been pulled in.

**Comment by noahtallen** [comment]
Hm. What if:

1. We only download updates to sources if the config changes (e.g. same logic we have now)
2. We add a flag to download updates to sources (like `wp-env start --update`) which also re-configures wp.

**Comment by epiqueras** [comment]
That makes more sense. People might not want to update sometimes too.

**Comment by noahtallen** [comment]
Makes sense. Added in c36e22d.

File: packages/env/lib/commands/start.js [hunk 2]
```diff
@@ -50,24 +66,20 @@ module.exports = async function start( { spinner, debug } ) {
...
+		config: config.dockerComposeConfigPath,
+		log: config.debug,
+	} );
 
-	await setupWordPressDirectories( config );
+	if ( shouldConfigureWp ) {
+		spinner.text = 'Downloading sources.';
+		await downloadSources( config, spinner );
+		await setupWordPressDirectories( config );
+	}
```
Comment (epiqueras) [inline packages/env/lib/commands/start.js:82]: Can't we still parallelize this?

File: packages/env/lib/commands/start.js [hunk 3]
```diff
@@ -76,37 +88,47 @@ module.exports = async function start( { spinner, debug } ) {
...
 
-	try {
-		await checkDatabaseConnection( config );
-	} catch ( error ) {
-		// Wait 30 seconds for MySQL to accept connections.
-		await retry( () => checkDatabaseConnection( config ), {
-			times: 30,
-			delay: 1000,
+		// Set the cache key once everything has been configured.
+		await setCache( CONFIG_CACHE_KEY, md5( config ), {
```
Comment (jeyip) [inline packages/env/lib/commands/start.js:127]: `Non-blocking`:

On [L54](https://github.com/WordPress/gutenberg/pull/23809/files#diff-03588f5a7a029851c0d4bde5d961748fR54) we calculate the md5 hash to evaluate whether or not the cache changed. It looks like we duplicate work here by recalculating the hash.

Could we use a temp variable?

**Review by jeyip — COMMENTED** [review]
Spent a good amount of time wrapping my head around the `wp-env` package. Definitely not an expert, but the caching logic looks reasonable. 

I can also confirm that:
- the cache is busted when `.wp-env.json` options change
- the performance is blazing fast 🔥

I'm having trouble verifying that adding the --update flag busts the cache though.

**Comment by noahtallen** [comment]
> I'm having trouble verifying that adding the --update flag busts the cache though.

if it takes > 2 seconds then it should be re-configuring wp. In Gutenberg, you can also verify by noting that it shows "downloading WordPress" for a brief moment before continuing (if it is already up to date)

File: packages/env/lib/commands/start.js (hunk 2 above)
Comment (noahtallen) [reply packages/env/lib/commands/start.js:82]: done

File: packages/env/lib/commands/start.js (hunk 3 above)
Comment (noahtallen) [reply packages/env/lib/commands/start.js]: sure! done :)

**Review by jeyip — COMMENTED** [review]
>if it takes > 2 seconds then it should be re-configuring wp. In Gutenberg, you can also verify by noting that it shows "downloading WordPress" for a brief moment before continuing (if it is already up to date)

My mistake. I should clarify. When I said "I'm having trouble verifying", I meant that I wasn't able to force the longer load times. Passing the --update flag didn't lead to expected behavior.

**Comment by noahtallen** [comment]
@jeyip oh, interesting! Here's a gif of what it does for me:

**Comment by jeyip** [comment]
@noahtallen Let me try again. I'll keep you updated.

**Comment by jeyip** [comment]
- `npm run wp-env start --update` looks like it references the cache
- `npx wp-env start --update` looks like it forces the reconfiguration

npm run
npx

**Comment by noahtallen** [comment]
🤔 🤔 🤔 🤔 🤔 🤔 🤔 🤔 🤔

**Comment by noahtallen** [comment]
That's confusing, since both should ultimately be accessing `./packages/env/bin/wp-env`. And it's even more confusing because both are clearly picking up on _some_ of the cache changes

**Comment by noahtallen** [comment]
@jeyip I figured it out actually 😅 If you run `npm run wp-env --update`, the update flag is not passed to wp-env. You have to run `npm run wp-env -- --update`

Screen Shot 2020-07-13 at 11 19 10 AM

**Comment by jeyip** [comment]
Oh I see! This seems like user error then, since [npm documents this pretty clearly](https://docs.npmjs.com/cli/run-script). Everything else looks good to me 🎉

**Review by jeyip — APPROVED** [review]

File: packages/env/lib/cache.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+	);
+}
+
+/**
+ * This function retrieves the cache associated with the given key from the file.
+ * Returns undefined if the key does not exist or if the cache file has not been
+ * created yet.
+ *
+ * @param {string}           key     The unique identifier for the cache value.
+ * @param {WPEnvCacheOptions} options Parsing options
```
Comment (epiqueras) [inline packages/env/lib/cache.js:57]: Alignment

File: packages/env/lib/cache.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+}
+
+/**
+ * This function persists the given cache value to the cache file. It creates the
+ * file if it does not exist yet, and overwrites the existing cache value for the
+ * given key if it already exists.
+ *
+ * @param {string}           key     A unique identifier for the cache.
+ * @param {any}              value   The value to persist.
+ * @param {WPEnvCacheOptions} options Parsing options
```
Comment (epiqueras) [inline packages/env/lib/cache.js:39]: Alignment

File: packages/env/lib/cache.js [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+const CACHE_FILE_NAME = 'wp-env-cache.json';
+
+/**
+ * This function can be used to compare a possible new cache value against an
+ * existing cache value. For example, we can use this to check if the configuration
+ * has changed in a new run of wp-env start.
+ *
+ * @param {string}           key     A unique identifier for the cache.
+ * @param {any}              value   The value to check against the existing cache.
+ * @param {WPEnvCacheOptions} options Parsing options
```
Comment (epiqueras) [inline packages/env/lib/cache.js:23]: Alignment

**Review by epiqueras — APPROVED** [review]

File: packages/env/lib/cache.js (hunk 4 above)
Comment (noahtallen) [reply packages/env/lib/cache.js:57]: 👍 makes sense. when i renamed from `meta` to `cache` it knocked it out of alignment
